beacon-chain/cache/depositsnapshot/deposit_tree.go

```diff
@@ -0,0 +1,99 @@
+package depositsnapshot
+
+import (
+	"crypto/sha256"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v3/encoding/bytesutil"
+	eth "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+)
+
+var (
+	ErrEmptyExecutionBlock = errors.New("empty execution block")
+	ErrInvalidSnapshotRoot = errors.New("snapshot root is invalid")
+)
+
+// DepositTree is the Merkle tree representation of deposits.
+type DepositTree struct {
+	tree                    MerkleTreeNode
+	mixInLength             uint64 // number of deposits in the tree, reference implementation calls this mix_in_length.
+	finalizedExecutionBlock ExecutionBlock
+}
+
+type ExecutionBlock struct {
+	Hash  [32]byte
+	Depth uint64
+}
+
+// New creates an empty deposit tree.
+func New() *DepositTree {
+	var leaves [][32]byte
+	merkle := create(leaves, DepositContractDepth)
+	return &DepositTree{
+		tree:                    merkle,
+		mixInLength:             0,
+		finalizedExecutionBlock: ExecutionBlock{},
+	}
+}
+
+func (d *DepositTree) getSnapshot() (DepositTreeSnapshot, error) {
+	if d.finalizedExecutionBlock == (ExecutionBlock{}) {
+		return DepositTreeSnapshot{}, ErrEmptyExecutionBlock
+	}
+	var finalized [][32]byte
+	depositCount, _ := d.tree.GetFinalized(finalized)
+	return fromTreeParts(finalized, depositCount, d.finalizedExecutionBlock), nil
+}
+
+func fromSnapshot(snapshot DepositTreeSnapshot) (DepositTree, error) {
+	if snapshot.depositRoot != snapshot.CalculateRoot() {
+		return DepositTree{}, ErrInvalidSnapshotRoot
+	}
+	finalizedExecutionBlock := ExecutionBlock{
+		Hash:  snapshot.executionBlockHash,
+		Depth: snapshot.executionBlockHeight,
+	}
+	tree := fromSnapshotParts(snapshot.finalized, snapshot.depositCount, DepositContractDepth)
+	return DepositTree{
+		tree:                    tree,
+		mixInLength:             snapshot.depositCount,
+		finalizedExecutionBlock: finalizedExecutionBlock,
+	}, nil
+}
+
+func (d *DepositTree) finalize(eth1data eth.Eth1Data, executionBlockHeight uint64) {
```

Comment:
@prestonvanloon which `Eth1Data` should I be using here?

Reply:
I used `eth "github.com/prysmaticlabs/prysm/v3/proto/eth/v1"` since it made more sense to me but there is also `eth "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"` so wanted to double check.

Reply:
Ah -- Those are essentially the same thing. We had to duplicate it for some compatibility reasons with ssz-gen tooling